Refactored iOS LlmInference and updated the app to use Sessions API

priankakariat opened the pull request:
### Description
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context.
1. Refactored iOS LlmInference 
2. Updated the app to use sessions API.
3. Updated the UI and fixed a few UI issues.
Fixes # (issue)
1. Fixed a few UI issues.
2. DIsabled message sending during response generation.
3. Added a progress view for model loading.
4. Other minor issues
### Checklist
Please ensure the following items are complete before submitting a pull request:

- [*] My code follows the code style of the project.
- [*] I have updated the documentation (if applicable).
- [ ] I have added tests to cover my changes.

### Type of Change
Please check the relevant option below:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation update (non-breaking change which updates documentation)
- [ *] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Screenshots
If applicable, please add screenshots to help explain your changes.

### Additional Notes
Add any additional information or context about the pull request here.

yishuangP commented on examples/llm_inference/ios/InferenceExample/View Models/ConversationViewModel.swift:
```diff
@@ -0,0 +1,326 @@
+    case loadingModel
+    case promptSubmitted
+    case streamingResponse
+    case error
```
Can we make this case take in an InferenceError? For example, case failed(error: InferenceError). Then we can just keep the currentState property and remove criticalError and createChatError?

yishuangP commented on examples/llm_inference/ios/InferenceExample/View Models/ConversationViewModel.swift:
```diff
@@ -0,0 +1,326 @@
+  func sendMessage(_ text: String) {
+    Task {
+      await internalSendMessage(text)
+      currentState = .done
```
nit: remove this? `internalSendMessage` already update currentState.

priankakariat commented on examples/llm_inference/ios/README.md:
```diff
@@ -1,5 +1,6 @@
+=====
 
 1. Run `pod install`.
-1. Download the `gemma-2b-it-cpu-int4.bin` and `gemma-2b-it-gpu-int4.bin` model from [the documentation](https://developers.google.com/mediapipe/solutions/genai/llm_inference#models) into this directory.
```
Rolled back the changes. It was a merge issue on my end.

priankakariat commented on examples/llm_inference/ios/InferenceExample/InferenceExample.entitlements:
```diff
@@ -2,8 +2,6 @@
 <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
 <plist version="1.0">
 <dict>
-    <key>com.apple.developer.kernel.extended-virtual-addressing</key>
```
Rolled back the update. Caused due to a merge issue on my end

priankakariat commented on examples/llm_inference/ios/InferenceExample/View Models/ConversationViewModel.swift:
```diff
@@ -0,0 +1,326 @@
+  }
+
+  func update(participant: ChatMessage.Participant) {
+    guard chatMessage.participant != .user && participant != .user else {
```
Updating the participant is only allowed if the chat message is of any  `.system` type and the it can only be updated to one of the other `.system`. This is to support UI differentiation in model messages for "Thinking", "Generating", completed streaming etc. A user message type should be uneditable.

priankakariat commented on examples/llm_inference/ios/InferenceExample/InferenceExampleApp.swift:
```diff
@@ -16,6 +16,11 @@ import SwiftUI
+//  init() {
+////    UILabel.appearance(whenContainedInInstancesOf: [UINavigationBar.self])
+////      .adjustsFontSizeToFitWidth = true
+//  }
```
I have removed this code entirely. I was trying to make the title scale according to device's screen size. Instead I have set the navigation bar title display mode to inline. That scales according to device size and looks more neater than the default which just sits kinda below the navigation bar buttons. This code was necessary for scaling the title according to screen size if the default mode was used.

priankakariat commented on examples/llm_inference/ios/InferenceExample/View Models/ConversationViewModel.swift:
```diff
@@ -0,0 +1,326 @@
+}
+
+@MainActor
+class MessageViewModel: ObservableObject, Identifiable {
```
Moved the message view model to a new file. 

Since an existing model message UI can get updated based on growing text, whether there is a delay in generating first token, if its an error etc, it was easier to propogate these updates via a separate VM to sync it with the UI. 

Chat messages are structs, so updating existing messages in the array doesn't look very elegant in code at times. 
Also, more functionality might get added based on the new features.

I tried changing the `ChatMessage` to a class. I haven't got this to work well as an observable object for the UI updates. If I can fix that implementation, we can manybe move it back to the conversation VM.
I am focusing on adding DeepSeek support. I'll revisit this later.

schmidt-sebastian commented:
Hello! I did not review the code (sorry...) but only the functionality of the app. It is so much better already!

Here are some things I noticed:
- When I get an error that the model file is not found, the app still opens the chat window rather than going back to the model selection.
- The app gets killed when I switch from one inference session to the next. This could be the case if we don't close the inference engine before we initialize a new one.
- We should add the Gemma prompt prefixes: https://github.com/google-ai-edge/mediapipe-samples/blob/main/examples/llm_inference/android/app/src/main/java/com/google/mediapipe/examples/llminference/ChatUiState.kt#L122

Thank you!

priankakariat commented:
> Hello! I did not review the code (sorry...) but only the functionality of the app. It is so much better already!
> 
> Here are some things I noticed:
> 
> * When I get an error that the model file is not found, the app still opens the chat window rather than going back to the model selection.
> * The app gets killed when I switch from one inference session to the next. This could be the case if we don't close the inference engine before we initialize a new one.
> * We should add the Gemma prompt prefixes: https://github.com/google-ai-edge/mediapipe-samples/blob/main/examples/llm_inference/android/app/src/main/java/com/google/mediapipe/examples/llminference/ChatUiState.kt#L122
> 
> Thank you!

I have fixed the issues. 

The app was quitting due to memory pressure on the device. I was testing on the simulator, so didn't encounter this. SwiftUI's navigation stack's memory management is quite different from UIKit. I wasn't aware of this. The popped conversation view doesn't get cleared by the OS from memory when you go back to the model selection screen and select a different model.  Along with the old view and it's VM, the LlmInference object also remains in memory.
I don't know if its supposed to be a bug or a feature of SwiftUI. 

I have slightly refactored model initialization to adjust to this behaviour. 

The session gets closed correctly. I incorporated this into the library. When the session is deinitialized by the OS, the C++ session will also be deleted. Users doesn't have to explicitly call a close in iOS. Also, haven't exposed a close because of this.

schmidt-sebastian commented:
Thank you for the updates. Two more things we should fix and then we can merge it:

- When the model file is not found, the model selection view pops up again. But the first time this happens, the dialog box pops up for a second or so.
- The prompt formatting does not seem to be invoked. I changed the following to get it to work:


```
private func internalSendMessage(_ text: String) async {
...

    defer { systemViewModel.closeSystemMessage() }

    do {
      let prompt = formatPrompt(text:text)
      let responseStream = try await chat.sendMessage(prompt)

      await updateSystemViewModel(systemViewModel, responseStream: responseStream)
    } catch {
      /// systemViewModel is closed in a defer before exiting this scope. Any errors are handled during close.
    }

```
